sdk/src/test/java/com/mapbox/maps/MapControllerTest.kt

```diff
@@ -8,250 +8,310 @@ import com.mapbox.maps.loader.MapboxMapStaticInitializer
-    every { MapProvider.getMapPluginRegistry(any(), any(), any()) } returns pluginRegistry
-    every { mapboxMap.isStyleLoadInitiated } returns false
-    every { renderer.renderThread.handlerThread.handler } returns mockk()
-    mapController =
-      MapController(
-        renderer,
-        nativeObserver,
-        mapInitOptions,
-        nativeMap,
-        mapboxMap,
-        pluginRegistry,
-        mockk()
-      )
+    every { mockContext.packageName } returns "com.mapbox.maps"
+
+    testMapController = MapController(
+      mockRenderer,
+      mockNativeObserver,
+      mockMapInitOptions,
+      mockNativeMap,
+      mockMapboxMap,
+      mockPluginRegistry,
+      mockOnStyleDataLoadedListener
+    )
+  }
+
+  @After
+  fun shutDown() {
+    unmockkAll()
   }
 
   @Test
   fun onStart() {
-    mapController.onStart()
-    verify { renderer.onStart() }
-    verify { pluginRegistry.onStart() }
-    verify { mapboxMap.loadStyleUri(Style.MAPBOX_STREETS) }
+    every { mockPluginRegistry.onStart() } just Runs
+    every { mockMapboxMap.loadStyleUri(Style.MAPBOX_STREETS) } just Runs
+    every { mockNativeObserver.addOnCameraChangeListener(any()) } just Runs
+    every { mockNativeObserver.addOnStyleDataLoadedListener(any()) } just Runs
+    every { mockRenderer.onStart() } just Runs
+    every { mockMapboxMap.isStyleLoadInitiated } returns false
+    every { mockMapInitOptions.styleUri } answers { Style.MAPBOX_STREETS }
+
+    testMapController.onStart()
+
+    verifySequence {
+      mockNativeObserver.addOnCameraChangeListener(any())
+      mockNativeObserver.addOnStyleDataLoadedListener(any())
+      mockRenderer.onStart()
+      mockMapboxMap.isStyleLoadInitiated
+      mockMapboxMap.loadStyleUri(Style.MAPBOX_STREETS)
+      mockPluginRegistry.onStart()
+    }
   }
 
   @Test
   fun onStartWithStyleLoaded() {
-    every { mapboxMap.isStyleLoadInitiated } returns true
-    mapController.onStart()
-    verify(exactly = 0) { mapboxMap.loadStyleUri(Style.MAPBOX_STREETS) }
+    every { mockMapboxMap.isStyleLoadInitiated } returns true
+    every { mockPluginRegistry.onStart() } just Runs
+    every { mockRenderer.onStart() } just Runs
+    every { mockNativeObserver.addOnCameraChangeListener(any()) } just Runs
+    every { mockNativeObserver.addOnStyleDataLoadedListener(any()) } just Runs
+
+    testMapController.onStart()
+
+    verify(exactly = 0) { mockMapboxMap.loadStyleUri(Style.MAPBOX_STREETS) }
   }
 
   @Test
   fun onStop() {
-    mapController.onStart()
-    mapController.onStop()
-    verify { renderer.onStop() }
-    verify { pluginRegistry.onStop() }
+    every { mockPluginRegistry.onStop() } just Runs
+    every { mockRenderer.onStop() } just Runs
+    every { mockNativeObserver.removeOnCameraChangeListener(any()) } just Runs
+    every { mockNativeObserver.removeOnStyleDataLoadedListener(any()) } just Runs
+
+    testMapController.simulateStartedState()
```

Comment:
can we use `onStart` directly instead of using `simulateStartState`?

Reply:
+1 for that and remove `simulateStartState` at all. imho there's no possibility in reality to set `lifecycleState = LifecycleState.STATE_STARTED` and __not__ call `onStart` at the same time

Reply:
> I disagree with relaxed, I'm trying to follow Arrange-Action-Assert pattern to write a good tests.

Is this comment related to https://github.com/mapbox/mapbox-maps-android/pull/680#discussion_r721203936 actually?